### PR TITLE
add ability to exclude daemon threads

### PR DESCRIPTION
**Pull request by sv3ndk**
Hi,

I used your tool yesterday to troubleshoot a flakky test suite that was randomly hanging due to leaked threads. It's quite handy, thanks for sharing it.

In my situation, I primarily wanted to focus on non-daemon leaked threads since they were causing the biggest issue.

This PR adds the `threadleak_exclude_daemon_threads` option in the `.ini` file: when true, leaked daemon threads are not causing a test failure.

I also extended the tox config to python 3.11 and 3.12.

I hope you find this useful!

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -108,6 +108,17 @@ regex to match excluded thread names:
...
 
+By default, all leaked threads are reported. If you want to ignore leaked
+daemon threads, you can exclude them:
+
+
+.. code-block:: ini
+
+    [pytest]
+    threadleak = True
+    threadleak_exclude_daemon_threads = True
```
Comment (nirs) [inline README.rst]: This may be useful for some cases, but why you cannot ignore the threads by name? Your system starts daemon threads without useful names? It feels that the right place to fix this issue is by naming your threads in consistent way.

Also the name is little bit too long, threadleak is a plugin for detecting thread leaks, so we can ommit the _threads suffix and have threadleak_exclude_daemons.

File: pytest_threadleak.py [hunk 2]
```diff
@@ -25,6 +25,11 @@ def pytest_addoption(parser):
     parser.addini(
         'threadleak_exclude',
         'Regex of thread names to exclude')
+    parser.addini(
+        'threadleak_exclude_daemon_threads',
```
Comment (nirs) [inline pytest_threadleak.py]: threadleak_exclude_daemons?

File: pytest_threadleak.py [hunk 3]
```diff
@@ -38,11 +43,13 @@ def pytest_configure(config):
 def pytest_runtest_call(item):
     start_threads = None
     exclude_regex = item.config.getini("threadleak_exclude")
+    exclude_daemon = item.config.getini("threadleak_exclude_daemon_threads")
```
Comment (nirs) [inline pytest_threadleak.py]: exclude_daemons?

File: pytest_threadleak.py [hunk 4]
```diff
@@ -75,11 +82,16 @@ def check_marker(marker):
             "Unexpected threadleak kwargs: {}".format(marker.kwargs))
 
 
-def current_threads(exclude_regex=None):
+def current_threads(exclude_regex=None, exclude_daemon=False):
```
Comment (nirs) [inline pytest_threadleak.py]: exclude_daemons=False?

File: test_threadleak.py [hunk 5]
```diff
@@ -3,6 +3,7 @@
 # SPDX-License-Identifier: MIT
 
 import pytest
+from _pytest.config import ExitCode
```
Comment (nirs) [inline test_threadleak.py]: Why import private pytest names? This is fragile.

File: test_threadleak.py [hunk 6]
```diff
@@ -64,7 +73,7 @@ def test_leak_enabled_option(testdir):
         '*::test_leak FAILED*',
         '*Failed: Test leaked *leaked-thread-0*leaked-thread-1*',
     ])
-    assert result.ret == 1
+    assert result.ret == ExitCode.TESTS_FAILED
```
Comment (nirs) [inline test_threadleak.py]: I don't see much value in this change. It can make sense if the enums are public. Otherwise using 0 and 1 is fine. If you really want to use enums and there is no way to import them from pytest without accessing _private packages, you can define them in this file.

This change is also not related to the new feature, so it is best done in another PR.

File: test_threadleak.py [hunk 7]
```diff
@@ -110,23 +119,28 @@ def test_leak_disabled_marker(testdir):
...
 def test_leak_enabled_class_marker(testdir):
     testdir.makepyfile(make_source(
         class_marker="@pytest.mark.threadleak"))
     result = testdir.runpytest('-v')
-    result.stdout.fnmatch_lines(['*::test_leak FAILED*'])
-    assert result.ret == 1
+    print(f"""--]]]]]]]]]]]]------------
+    {result.stdout.str()}
+    ----++++++++++++++++------------
+    """)
```
Comment (nirs) [inline test_threadleak.py]: I see that you removed this debug print in a later commit, but this makes review harder and requires squashing the commits to keep clean history. It would be useful if you can squash your commits before the next review to separate logical changes.

File: tox.ini [hunk 8]
```diff
@@ -4,7 +4,7 @@
 
 # For more information about tox, see https://tox.readthedocs.io/en/latest/
 [tox]
-envlist = py27,py33,py34,py35,py36,py37,py38,py39,py310,pypy,flake8
+envlist = py27,py33,py34,py35,py36,py37,py38,py39,py310,py311,py312,pypy,flake8
```
Comment (nirs) [inline tox.ini]: Adding newer python version is good, but we really need to get the test suite working first. Currently it fails with py2.7 and py < 3.8, probably because the versions are not availlable now in github actions.

It will be better to start with PR with this commit, and remove the missing python versions, or send another trivial PR wit these changes.

File: README.rst
Comment (nirs) [inline README.rst:1]: This cleanups commit should be squashed into the first one.

**Review by nirs — COMMENTED** [review]
@sv3ndk thanks for your PR!

The new option makes sense if you have a system that you cannot modify to get consistent thread names that can be excluded by name. Maybe you can add more context on why this is not possible in your case?

It would be nice if you can separate the unrelated changes (ExitCode eunm) and python versions and first get the test suite running again.

File: tox.ini (hunk 8 above)
Comment (sv3ndk) [reply tox.ini]: Ok, let's try this: https://github.com/nirs/pytest-threadleak/pull/13

**Comment by nirs** [comment]
@sv3ndk I'm adding black format check in #15. Reformatting your branch with black before you rebase on master should make it easy to rebase.

**Comment by sv3ndk** [comment]
> Maybe you can add more context on why this is not possible in your case?

Hi,

I was troubleshooting an existing application, which has its own conventions for naming threads. After a recent change, the pytest test suite was never terminating even though all tests were successful, suggesting some thread had leaked from one of our 1000+ unit tests, with no indication which one nor from which test. Temporarily activating `pytest-threadleak` helped to localize the issue since it forced the relevant test to fail immediately with a clear diagnostic log.

**Comment by sv3ndk** [comment]
> @sv3ndk I'm adding black format check in #15. Reformatting your branch with black before you rebase on master should make it easy to rebase.

Ok, I've not yet included black in this PR to keep it easy to read, I'll rebase once again when https://github.com/nirs/pytest-threadleak/pull/15 is merged

**Comment by nirs** [comment]
> Ok, I've not yet included black in this PR to keep it easy to read, I'll rebase once again when #15 is merged

Merged now.

**Comment by nirs** [comment]
@sv3ndk thanks for the PR!
